packages/opentelemetry-instrumentation-ws/src/index.ts

```diff
@@ -430,3 +435,49 @@ export class WSInstrumentation extends InstrumentationBase<WS> {
     };
   };
 }
+
+class TimeStampedSpan implements Span {
+  constructor(private readonly span: Span) {}
+  spanContext(): SpanContext {
+    return this.span.spanContext();
+  }
+  setAttribute(key: string, value: SpanAttributeValue): this {
+    this.span.setAttribute(key, value);
+    return this;
+  }
+  setAttributes(attributes: SpanAttributes): this {
+    this.span.setAttributes(attributes);
+    return this;
+  }
+  addEvent(name: string, attributesOrStartTime?: SpanAttributes | TimeInput, startTime?: TimeInput): this {
+    this.span.addEvent(name, attributesOrStartTime, startTime);
+    return this;
+  }
+  setStatus(status: SpanStatus): this {
+    this.span.setStatus(status);
+    return this;
+  }
+  updateName(name: string): this {
+    this.span.updateName(name);
+    return this;
+  }
+  isRecording(): boolean {
+    return this.span.isRecording();
+  }
+  recordException(exception: Exception, time?: TimeInput): void {
+    this.span.recordException(exception, time);
+  }
+  private _endTime: TimeInput | undefined;
+
+  set endTime(endTime: TimeInput) {
+    this._endTime = endTime;
+  }
+
+  end(endTime?: TimeInput): void {
+    if (this._endTime) {
+      this.span.end(this._endTime);
```

Comment:
I think this accomplishes what you initially intended, to mark the WS upgrade as closed once its work is done and before all of the work on the WS is done. We're saving the actual end of that work and backdating the span end essentially. Let me know what you think.